src/nbl/asset/utils/ISPIRVDebloater.cpp

```diff
@@ -0,0 +1,115 @@
+#include "nbl/asset/utils/ISPIRVDebloater.h"
+#include "nbl/asset/utils/ISPIRVOptimizer.h"
+#include "nbl_spirv_cross/spirv.hpp"
+
+#include "nbl/core/declarations.h"
+#include "nbl/system/ILogger.h"
+
+
+using namespace nbl::asset;
+
+ISPIRVDebloater::ISPIRVDebloater()
+{
+    constexpr auto optimizationPasses = std::array{
+        ISPIRVOptimizer::EOP_ELIM_DEAD_FUNCTIONS,
+        ISPIRVOptimizer::EOP_ELIM_DEAD_VARIABLES,
+        ISPIRVOptimizer::EOP_ELIM_DEAD_CONSTANTS,
+        // TODO: remove unused type. no dedicated flag for type. maybe need agressive dce to remove type.
+        ISPIRVOptimizer::EOP_TRIM_CAPABILITIES,
+    };
+    m_optimizer = core::make_smart_refctd_ptr<ISPIRVOptimizer>(std::span(optimizationPasses));
+}
+
+ISPIRVDebloater::Result ISPIRVDebloater::debloat(const  ICPUBuffer* spirvBuffer, std::span<const EntryPoint> entryPoints, system::logger_opt_ptr logger) const
+{
+    const auto* spirv = static_cast<const uint32_t*>(spirvBuffer->getPointer());
+    const auto spirvDwordCount = spirvBuffer->getSize() / 4;
+
+    auto getHlslShaderStage = [](spv::ExecutionModel executionModel) -> hlsl::ShaderStage
+      {
+        switch (executionModel)
+        {
+            case  spv::ExecutionModelVertex : return hlsl::ESS_VERTEX;
+            case  spv::ExecutionModelTessellationControl : return hlsl::ESS_TESSELLATION_CONTROL;
+            case  spv::ExecutionModelTessellationEvaluation : return hlsl::ESS_TESSELLATION_EVALUATION;
+            case  spv::ExecutionModelGeometry : return hlsl::ESS_GEOMETRY;
+            case  spv::ExecutionModelFragment : return hlsl::ESS_FRAGMENT;
+            case  spv::ExecutionModelGLCompute : return hlsl::ESS_COMPUTE;
+            case  spv::ExecutionModelTaskEXT : return hlsl::ESS_TASK;
+            case  spv::ExecutionModelMeshEXT: return hlsl::ESS_MESH;
+            case  spv::ExecutionModelRayGenerationKHR : return hlsl::ESS_RAYGEN;
+            case  spv::ExecutionModelAnyHitKHR : return hlsl::ESS_ANY_HIT;
+            case  spv::ExecutionModelClosestHitKHR : return hlsl::ESS_CLOSEST_HIT;
+            case  spv::ExecutionModelMissKHR : return hlsl::ESS_MISS;
+            case  spv::ExecutionModelIntersectionKHR : return hlsl::ESS_INTERSECTION;
+            case  spv::ExecutionModelCallableKHR : return hlsl::ESS_CALLABLE;
+            default:
+            {
+                assert(false);
```

Comment:
NEVER assert to handle a regular user error, we dont want to crash in debug mode because someone misused the engine

Use the logger instead.